expr/expr.go

```diff
@@ -245,542 +274,371 @@ func floatToUint64(f float64) (uint64, bool) {
+	case ">":
+		c.convert = func(v int) bool { return v > 0 }
+	case ">=":
+		c.convert = func(v int) bool { return v >= 0 }
+	default:
+		return nil, fmt.Errorf("unknown comparison operator: %s", operator)
+	}
+	return c, nil
 }
 
-func compileCompareRelative(lhsFunc, rhsFunc NativeEvaluator, operator string) (NativeEvaluator, error) {
-	return func(rec *zng.Record) (zngnative.Value, error) {
-		lhs, err := lhsFunc(rec)
-		if err != nil {
-			return zngnative.Value{}, err
-		}
-		rhs, err := rhsFunc(rec)
-		if err != nil {
-			return zngnative.Value{}, err
-		}
+func (c *Compare) result(result int) zng.Value {
+	if c.convert(result) {
+		return zng.True
+	}
+	return zng.False
+}
 
-		// holds
-		//   <0 if lhs < rhs
-		//    0 if lhs == rhs
-		//   >0 if lhs > rhs
-		var result int
-		switch lhs.Type.ID() {
-		case zng.IdInt16, zng.IdInt32, zng.IdInt64, zng.IdTime, zng.IdDuration:
-			lv := lhs.Value.(int64)
-			var rv int64
-
-			switch rhs.Type.ID() {
-			case zng.IdByte, zng.IdUint16, zng.IdUint32, zng.IdUint64, zng.IdPort:
-				if (lhs.Type.ID() == zng.IdTime || lhs.Type.ID() == zng.IdDuration) && rhs.Type.ID() == zng.IdPort {
-					return zngnative.Value{}, ErrIncompatibleTypes
-				}
-
-				// signed/unsigned comparison
-				runsigned := rhs.Value.(uint64)
-				if lv < 0 {
-					result = -1
-					break
-				} else if runsigned > math.MaxInt32 {
-					result = 1
-					break
-				}
-				rv = int64(runsigned)
-
-			case zng.IdInt16, zng.IdInt32, zng.IdInt64, zng.IdTime, zng.IdDuration:
-				if (lhs.Type.ID() == zng.IdTime && rhs.Type.ID() == zng.IdDuration) || (lhs.Type.ID() == zng.IdDuration && rhs.Type.ID() == zng.IdTime) {
-					return zngnative.Value{}, ErrIncompatibleTypes
-				}
-				rv = rhs.Value.(int64)
-			case zng.IdFloat64:
-				lf := float64(lv)
-				rf := rhs.Value.(float64)
-				if lf < rf {
-					result = -1
-				} else if lf == rf {
-					result = 0
-				} else {
-					result = 1
-				}
-
-			default:
-				return zngnative.Value{}, ErrIncompatibleTypes
+func (c *Compare) Eval(rec *zng.Record) (zng.Value, error) {
+	lhs, err := c.lhs.Eval(rec)
+	if err != nil {
+		return zng.Value{}, err
+	}
+	rhs, err := c.rhs.Eval(rec)
+	if err != nil {
+		return zng.Value{}, err
+	}
+	id, err := c.vals.coerce(lhs, rhs)
+	if err != nil {
+		// If coercion fails due to overflow, then we know there is a
+		// mixed signed and unsigned situation and either the unsigned
+		// value couldn't be converted to an int64 because it was too big,
+		// or the signed value couldn't be converted to a uint64 because
+		// it was negative.  In either case, the unsigned value is bigger
+		// than the signed value.
+		if err == ErrOverflow {
+			result := 1
+			if zng.IsSigned(lhs.Type.ID()) {
+				result = -1
 			}
-			if lv < rv {
+			return c.result(result), nil
+		}
+		return zng.False, err
+	}
+	var result int
+	if !c.vals.equal() {
+		if zng.IsFloat(id) {
+			v1, _ := zng.DecodeFloat64(c.vals.a)
+			v2, _ := zng.DecodeFloat64(c.vals.b)
+			if v1 < v2 {
```

Comment:
nit: Maybe keep this section as a `switch` statement:
```golang
		switch {
		case zng.IsFloat(id):
			v1, _ := zng.DecodeFloat64(c.vals.a)
			v2, _ := zng.DecodeFloat64(c.vals.b)
			if v1 < v2 {
				result = -1
			} else {
				result = 1
			}
		case zng.IsSigned(id):
			v1, _ := zng.DecodeInt(c.vals.a)
			v2, _ := zng.DecodeInt(c.vals.b)
			if v1 < v2 {
				result = -1
			} else {
				result = 1
			}
		case zng.IsNumber(id):
			v1, _ := zng.DecodeUint(c.vals.a)
			v2, _ := zng.DecodeUint(c.vals.b)
			if v1 < v2 {
				result = -1
			} else {
				result = 1
			}
...
```

Reply:
Good idea.  More readable as it exposes the structure more clearly.  I changed these if-else's throughout.